File: src/shrinker/dictionary.ts
import { Shrinkable } from '../Shrinkable'
import { shrinkableArray } from './array'
import { shrinkableTuple } from './tuple'

export interface Dictionary<T> {
    [Key: string]: T
}

/**
 * Creates a `Shrinkable<Dictionary<T>>` from an array of `[key, value]` Shrinkable pairs.
 *
 * Shrinks across three dimensions — matching `cppproptest`'s `shrinkMap` behaviour:
 *
 *  1. **Membership-wise** — removes key-value pairs using `shrinkFrontAndThenMid`
 *     (full 2^N subset exploration, not just rear-truncation).
 *  2. **Key shrinking** — each key is shrunk through its key generator's shrink tree
 *     (e.g. shorter strings, smaller integer keys).
 *  3. **Value shrinking** — each value is shrunk through its value generator's shrink tree.
 *
 * Key + value shrinking are combined via `shrinkableTuple(keyShr, valShr)` which
 * uses `concat` so both axes are accessible at every node of the membership tree.
 * Element-wise shrinking is wired with `concat` (not `andThen`) in `shrinkableArray`,
 * so pair shrinks fire at every membership node, not only at leaves.
 *
 * @param pairs   - Array of `[Shrinkable<string>, Shrinkable<T>]` entries.
 * @param minSize - Minimum number of key-value pairs after shrinking.
 */
export function shrinkableDictionary<T>(
    pairs: Array<[Shrinkable<string>, Shrinkable<T>]>,
    minSize: number
): Shrinkable<Dictionary<T>> {
    // Wrap each (keyShr, valShr) pair into a Shrinkable<[string, T]> using
    // shrinkableTuple.  The resulting Shrinkable carries key shrinks (position 0,
    // via the map chain from keyShr) and value shrinks (position 1, via concat)
    // at every node.
    const pairShrinkables = pairs.map(
        ([keyShr, valShr]) => shrinkableTuple(keyShr, valShr) as unknown as Shrinkable<[string, T]>
    )

    // Apply membership-wise + element-wise shrinking.
    // membershipWise = true  → shrinkFrontAndThenMid
    // elementWise    = true  → shrink key+value at every membership node (via concat)
    const shrinkableArr = shrinkableArray(pairShrinkables, minSize, true, true)

    // Convert Array<[string, T]> → Dictionary<T>.
    // If key shrinking produces a duplicate key, the later pair's value wins;
    // the dict will have fewer entries than before, but the property still
    // fails / holds correctly (a smaller dict is always a valid shrink candidate).
    return shrinkableArr.map(arr => {
        const dict: Dictionary<T> = {}
        for (const [key, value] of arr) {
            dict[key] = value
        }
        return dict
    })
}

File: dartproptest/lib/src/shrinker/dictionary.dart
import '../shrinkable.dart';
import 'integer.dart';
import 'advanced_array.dart';

/// Creates a dictionary from a list of key-value pairs.
Map<String, T> _createDictionary<T>(List<(String, T)> pairs) {
  final dict = <String, T>{};
  for (final pair in pairs) {
    dict[pair.$1] = pair.$2;
  }
  return dict;
}

/// Creates a shrinkable dictionary that shrinks by reducing the number of key-value pairs
/// and optionally by shrinking the individual values.
///
/// [dict] The dictionary of Shrinkable values.
/// [minSize] The minimum number of key-value pairs.
/// [elementWise] If true, applies element-wise shrinking to the values. Defaults to false.
/// Returns a Shrinkable dictionary.
Shrinkable<Map<String, T>> shrinkableDictionary<T>(
  Map<String, Shrinkable<T>> dict,
  int minSize, {
  bool elementWise = false,
}) {
  final size = dict.length;
  final rangeShrinkable =
      binarySearchShrinkable(size - minSize).map((s) => s + minSize);

  Shrinkable<List<(String, Shrinkable<T>)>> shrinkablePairs =
      rangeShrinkable.map((newSize) {
    if (newSize == 0) return <(String, Shrinkable<T>)>[];

    final pairs = <(String, Shrinkable<T>)>[];
    int i = 0;
    for (final entry in dict.entries) {
      if (i < newSize) {
        pairs.add((entry.key, entry.value));
        i++;
      } else {
        break;
      }
    }
    return pairs;
  });

  // Apply element-wise shrinking if enabled
  if (elementWise) {
    // Convert to list of Shrinkable<(String, T)> for element-wise shrinking
    final shrinkableValuePairs = shrinkablePairs.map((pairs) =>
        pairs.map((pair) => pair.$2.map((value) => (pair.$1, value))).toList());

    // Apply element-wise shrinking
    final elementWiseShrinks = shrinkableValuePairs
        .andThen((parent) => shrinkElementWise(parent, 0, 0));

    // Convert back to the original format
    shrinkablePairs = elementWiseShrinks.map((shrinks) => shrinks
        .map((shrink) => (shrink.value.$1, Shrinkable<T>(shrink.value.$2)))
        .toList());
  }

  return shrinkablePairs.map((pairs) => _createDictionary(
      pairs.map((pair) => (pair.$1, pair.$2.value)).toList()));
}
